src/x-terminal.js

```diff
@@ -251,7 +251,6 @@ class XTerminalSingleton {
 			if (cursor) {
 				const line = editor.lineTextForBufferRow(cursor.row)
 				selectedText = line
-				editor.moveDown(1)
```

Comment:
this is not the way it works in terminus see https://github.com/bus-stop/terminus#insert-selected-text

the last two at cursor position, the cursor shifts down to the next line, this way you can run any commands in sequence, by disabling this you remove that ability.

![generic](https://user-images.githubusercontent.com/31389848/76141468-e0a30e80-6064-11ea-9cbb-2f4503e5d78f.gif)

with is better

Reply:
I saw that but as I was testing it out I realized I didn't always want it to move to the next line. I figured pressing <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>i</kbd> just to insert and <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>i</kbd> + <kbd>Down Arrow</kbd> to insert and move to the next line works better.

Reply:
How about making it optional, default on? SO it suits both

Reply:
I feel like 1 extra key press isn't worth a setting. What is a situation where someone, can't or wouldn't want to, press the down key to go to the next line?

Reply:
Like I said, running commands in a sequence is faster than having to move down manually.

Reply:
You don't push <kbd>i</kbd> and <kbd>Down</kbd> at the same time. Left hand <kbd>Ctrl</kbd> + <kbd>Shift</kbd>, right hand <kbd>i</kbd> then <kbd>Down</kbd> then <kbd>i</kbd> then <kbd>Down</kbd> then <kbd>i</kbd>. I feel like it taking 500 ms to run each line instead of 200ms You might save 2 seconds or, if there are a bunch of lines, it would be faster to select all the lines and paste them into a terminal anyways.

Just out of curiosity, how often do you personally run commands this way and why?

Reply:
I cant say its everyday, but when I do its generally ensuring each step of a script is actually doing what its supposed to do before its let loose.

The ability to do this sequentially one handed is simple and effective.

Reply:
Can we move on to https://github.com/bus-stop/x-terminal/pull/23#issuecomment-596061358

Reply:
SO I take it you wont revert this?

Reply:
I'm really 50/50 on this issue. I want the behavior to be intuitive. If someone is moving over from terminus then they would expect the cursor to move down automatically but if they are using this feature for the first time it seems unexpected for it to automatically move down. (I definitely didn't expect it when I noticed it.)

Maybe @Kabouik can give a third perspective.

Does it make sense to move the cursor down automatically?

Reply:
And should it move the cursor down automatically on `insert-selected-text` or just `run-selected-text`?

Reply:
After trying it a bit, I wholeheartedly agree that moving down one line should at least be an option if disabled by default. I suppose the need is not obvious for all coding languages, but for the way I would use x-terminal in Atom, it would make a huge difference.

R scripts are just a succession of individual R commands. While I can often run them as chunks, or even all at once when the code just works, debugging a script most of the time requires running lines one by one, multiple times until the issue is identified, or when tweaking figure aesthetics. I wouldn't want to need both hands for such a repetitive task, since even though I don't need the mouse all the time, I often need it to place the cursor further away (faster than using keyboard when the new cursor position is far).

From my perspective, the ideal setting would be an option to toggle on or off the automatic line change when sending code, and an extra modifier to send a full chunk of text at once (chunks could be recognized using commented sections, or empty lines), like `ctrl-r` and `shift-ctrl-r`, respectively.

Thanks a lot to both of you for your awesome and rapid work by the way, I can now already use x-terminal in my main workflow and the side-by-side layout is by far the best for my use case:

![Screenshot from 2020-03-09 16:54:22](https://user-images.githubusercontent.com/7107523/76232208-af723c00-6226-11ea-837d-0d1c832a22b7.png)

(Well, as soon as I fix this package installation issue. :<)

Reply:
> And should it move the cursor down automatically on `insert-selected-text` or just `run-selected-text`?

I think that's kind of the difference I would expect between `run` and `insert`. I don't expect `insert` to execute the command, so it would be awkward if it moved my cursor without any action actually occurring in the terminal. After `run`, however, the command is executed immediately and I can no longer change it in the terminal, so there is no point for me to keep the cursor on the same command because that's already the past and I won't need to send it again*.

* Unless specifically debugging that line, I see your point. However, in that case I don't need as many repeated key presses so I can just move back up after trying once, or go directly in the terminal to find what is wrong and move up in the history with up arrow.

There could also just be different left-hand keybindings for all, if no conflicts, I think that would be the best of both worlds:
`alt-r`: run line/selection
`alt-f`: run line/selection and move down
`shift-alt-r`: insert line/selection
`shift-alt-f`: insert line/selection and move down

Reply:
>I think that's kind of the difference I would expect between run and insert. I don't expect insert to execute the command, so it would be awkward if it moved my cursor without any action actually occurring in the terminal

That is unaffected by this moving the cursor down automatically, at least it didnt here.

execute/run is the same

Reply:
So I think that is a +1 for moving to the next line automatically on `run` but not on `insert`. That is good enough for me to add it back. I don't think we need a setting to enable/disable it because it would be easier to press the up key on `run` or down key on `insert` then to change the setting when you want the other behavior.

Reply:
Thanks for adding it back. If there is no conflict, it could be useful to include both behaviors and distinguish them with a modifier key, as I described in the above lengthy post, every use case would be covered.

Alternatively, if sticking to only two keybinds, I believe `running + moving down` and `moving + not moving down` are probably both more often used than `inserting without running or moving`. I assume that if one is using a terminal in Atom, he/she likely is editing long commands in the editor anyway, not in the terminal, and therefore has no need for inserting without running. But if you disagree, I'm very happy with what you described in your post!

Reply:
I still think it should be optional because it actually isn't regular behavior.

Reply:
Alright, no problem, your suggestion would work for me too.

Reply:
Thats not what Im suggesting, Im suggesting to just make the cursor move down optional,

As for inserting text not being useful that depends. Still nothing to do with cursor moving down or not, those things are separate and should stay separate.

Reply:
We're misunderstanding each other: I know your suggestion was to make moving down an option, and I think that's a good idea. That would totally work for me.

Also just to disambiguate, I understood that `run` and `insert` have nothing to do with moving the cursor down or not. I was merely saying that the `insert` feature could perhaps be dropped plain and simple if deemed useless, which would make room for a second keybind, and this keybind could have been used for `run and move down` without making it an option. But forget this idea, you confirmed that `insert` is not useless, so it should not be dropped.

Reply:
We don't necessarily need more keybindings we could add commands without key bindings and let the user remap them if they want to.

| Command | Keybinding|
| --- | --- |
| `x-terminal:run-selected-text-and-move-down` | <kbd>ctrl</kbd> + <kbd>alt</kbd> + <kbd>r</kbd> |
| `x-terminal:run-selected-text` | none |
| `x-terminal:insert-selected-text-and-move-down` | none |
| `x-terminal:insert-selected-text` | <kbd>ctrl</kbd> + <kbd>alt</kbd> + <kbd>i</kbd> |

Reply:
All we need is the option to enable the move down cursor automatically, it has the the ability to be a per profile setting also in that case.

Re: Adding commands, With above option or without the user already should already be able to remap the existing to whatever they want,

I tried making this moving down cursor automatically optional (beacaue I believe its the simpler solution to both needs), but failed at the element.js entries :disappointed:  Im studying the code but as a newb at JS Im a bit over my depth. Still Ive learned a few things, at this rate, say by 2055 before Im competent at this hahaha, by then Atom will surely have a built in terminal that does it all for you, even the coding.

Reply:
> not a fan of such hidden knobs of sorts

They will still be in the command palette.

If someone actually wants to sometimes move down and sometimes not, I feel like changing a setting takes a lot more time than just pressing up or down arrow keys, seems counter productive.

If we do add a setting that is available in the profile than changing it in the settings panel will not change it in any profile other than the base profile. They will have to change it in every profile they created.

I still think the multiple commands gives us the best of all worlds.

Reply:
We are never going to agree with this. =(

A user will still have to assign keymaps manually, so hidden knobs since your proposal implies adding commands without keybinds, bringing the pallete up to find this takes longer for me, the command pallete is great for apps that dont have keybinds to given commands, 
Also having to do this repeatedly feel counterproductive and counter intuitive,

And in that scenario a user could assign keymaps to those commands, but would have to be outside the app in Atom, or modify source keymaps.json and add them and have to do that every time app updates.

Even if we add keymaps seems more work for the sake of toggle.

But so be it, I can easily run a modified build locally or use terminus for these things until I figure out how to add this myself.

Reply:
> bringing the pallete up to find this takes longer for me

But it only has to be done once.

> and have to do that every time app updates.

My keybindings stay when updating Atom.

Reply:
I edited my reply, made things simpler for you.

Reply:
99% of people who use the `run` or `insert` commands will either *always* want the cursor to move down or not, so both more commands and a toggle setting will work form them. But more commands will make it easier for that 1% that does use both ways.

Reply:
> terminus has this behavior optional fyi

I actually looked at terminus to see how it was implemented and it looks like it is not optional.

https://github.com/bus-stop/terminus/blob/6de77586843a122c24f028160b4192987e49b032/lib/view.js#L567

Terminus has the option to run or insert but not the option to move down or not.

Reply:
I'm not trying to say that you're wrong. I like this discussion. I'm trying to understand the best way to make this usable for most people because changing it later will be much harder than doing it right the first time.

Reply:
My personal preference would be all commands available, whether already mapped to keybindings or not, it's pretty close to what I suggested above. However, if you both end up agreeing on a toggle in the options, I like it as well. The most important is the "move down" feature is not dropped, because it really is a central feature in my workflow (and I assume other people working with R in Atom).

However, in the former case, it should be clearly stated somewhere in the readme or in the descriptions in settings that these extra commands exist. Even if it's in the command palette, there are many packages that do not document all their features in the settings and I keep discovering stuff years later when I accidentally type the right keyword in the command palette, I don't any one wants that.

Reply:
Lets see:

Uzitech wants to use cursor keys to move cursor down manually

Kabouik: is happy with commands in palette as along as thats documented

me, well We know what my stance is:

For toggle: (disabled per default for a compromise) **:note** this is existing default behavior in terminus, platformio-ide-terminal and other forks it, is also default behavior as it is also now in x-terminal (not accounting for the removal here) 

* No extra documentation is needed, only a demo, When available,  Should be self explanatory.
* No hidden configurations e.g adding keymaps manually once (though it also needs to be enabled once per profile).
* Does not interfere with any existing workflows ala Uzitech or Kabouik or mine.
* Available in command pallete and with keymaps (as they are now) no matter if you want to move cursor down automatically or not.
* Can be available to any profile including base profile, we can create profiles with or without toggle to suit any circumstances.

Toggles are design method to implement such options that change how things behave different from default.

Usability wise to really find out we would need to do A-B testing between methods with telemetry enabled, this is because we cant administer A-B testing to a large enough user base sample to determine which way people use the app)

Its worth noting that the only gain with a toggle is really not having to press one extra key between sequences (which can get quite long depending) and YMMV between users and workflows.

Reply:
Terminus doesn't have a way to prevent the cursor form automatically moving down.

Reply:
> * No extra documentation is needed, only a demo, When available, Should be self explanatory.

I feel like the same amount of documentation will be needed for both.

> * No hidden configurations e.g adding keymaps manually once (though it also needs to be enabled once per profile).

The configuration will be hidden in the settings/profile panel. It is just as easy to find new commands in the command palette as it is to find new settings in the settings panel.

> * Does not interfere with any existing workflows ala Uzitech or Kabouik or mine.

Neither way interferes if is always off or always on

> * Available in command pallete and with keymaps (as they are now) no matter if you want to move cursor down automatically or not.

It would be much easier with multiple commands to just assign different keys to each command that a user uses than toggle it every time the user wants to change the behavior.

> * Can be available to any profile including base profile, we can create profiles with or without toggle to suit any circumstances.

This is one thing where I feel a profile toggle would work better. If a user wants some terminals automatically move down and some terminals to not move down they would just have to set different profiles on each terminal, one profile with the toggle on and one profile with the toggle off. With the multiple command however the user could choose per line whether they want to move down or not instead of per teiminal.

Reply:
> I feel like the same amount of documentation will be needed for both.

Lets agree to disagree. We dont document all the existing options except what in readme, and most options are not there, so one more one less in docs...

BTW Im in no way going to document ALL the options either.

> The configuration will be hidden in the settings/profile panel. It is just as easy to find new commands in the command palette as it is to find new settings in the settings panel.

Earlier you said you like this discussion, I feel however we are running in circles.

>It would be much easier with multiple commands to just assign different keys to each command that a user uses than toggle it every time the user wants to change the behavior.

Again in circles.







> This is one thing where I feel a profile toggle would work better. If a user wants some terminals automatically move down and some terminals to not move down they would just have to set different profiles on each terminal, one profile with the toggle on and one profile with the toggle off. With the multiple command however the user could choose per line whether they want to move down or not instead of per teiminal.

Theres pros and cons to everything. And were running in circles.

>Neither way interferes if is always off or always on

It will interfere if its on, since you dont want it to move down and prefer to do it manually.

Reply:
Think of this situation:

A user wants the cursor to move down if the line is run but not if the line is inserted.

### With a toggle:

The user would have to open the settings/profile panel and change the toggle whenever they want to switch from run line to insert line.

### With multiple commands:

The user sets up keymaps once and just presses different keys when they want to run and move down or insert and not move down.

---

In this situation having multiple commands is **much** easier for the user. Is there a situation where having a toggle would be **much** easier than having multiple commands?

Reply:
The purpose of a toggle is to allow the user to change behavior. Like I said above, if a user wants to **always** move down or **always** not move down both methods would work equally well. But I feel like the multiple commands will work **much** better for anyone who wants to switch between the cursor moving down sometimes and not moving down sometimes.

Reply:
I cannot agree sorry. And since we wont agree, I again remove myself because I no longer find this discussion helpful. 

Im happy to either use terminus for this or a modified local version of x-terminal.

Reply:
I feel like we are arguing two different points with this conversation. You are arguing that initial setup will be easier with a toggle than with multiple commands, and I am arguing that **after** initial setup using the feature will be easier with multiple commands than a toggle.

I agree with you that initial setup will be harder with multiple commands (i.e. finding commands and remapping keys to commands that the user wants to use) than it will with a toggle (i.e. flipping a switch), but I believe that changing the behavior often is more important than the initial setup.

## With multiple commands

#### inital setup:

1. find commands.
2. edit keymaps.json so keys map to correct commands that the user wants.

#### switching behavior per line

1. press different keys based on what the user wants to do with that line.

## With Toggle

#### inital setup:

1. flip the toggle to the default position

#### switching behavior per line

1. open profile panel for the active terminal
2. flip toggle
3. press keys to `insert`, `run` line
4. open profile panel for the active terminal
5. flip toggle back to default

Reply:
> I cannot agree sorry.

What exactly don't you agree with? I do value your perspective a lot. I'm not trying to change your mind, I'm trying to change my mind. I want to make sure there isn't something I didn't consider.

> Im happy to either use terminus for this or a modified local version of x-terminal.

You keep acting like Terminus has a toggle for this but it doesn't.

Reply:
>switching behavior per line

5 steps is 3 steps to many, You forget that you have a default profile and you can have a defined profile with other settings.

In any case this is an edge case in my experience with how I use this in terminus, I never really had the need and if I did its not that difficult to undo.

> You keep acting like Terminus has a toggle for this but it doesn't.

it doesnt, I never said it did (the option to run the text is optional), moving cursor down its default behavior, its default behavior now in x-terminal and all the platformio base forks, It works, never had an issue with it. **And the only reason to propose a toggle is because you dont want the cursor to move down automatically and prefer to manually do it.**

I rather have it move to the next line automatically, always ON is fine with me. But Im not the only person whos going to use it. So the toggle would be an improvements I think over terminus design.

>What exactly don't you agree with? I do value your perspective a lot. I'm not trying to change your mind, I'm trying to change my mind. I want to make sure there isn't something I didn't consider.

If you want add the toggle and the hidden knobs I'de be fine with that too (however unlikely that seems to me). The opposite is true for me with just the hidden knobs implementation I dont find this way alone sufficiently simple.

That said, no one is going to use this feature night and day, I'm not., so hence the length of this discussion and the impassable impasse is disproportionate to the importance of the feature.

Im fine with whatever you do in the end, 

Im happy that x-terminal is getting attention and dev time its more important to have a well supported app like this in Atom rather than a half baked one. That said xterm.js is not exactly bug free and it offers me new problems I dot have with term.js.

You know the pros and cons of each proposal and as I cant contribute to the feature by adding the toggle, I really defer to your good graces. Its appreciated. =)

Reply:
> it doesnt, I never said it did

https://github.com/bus-stop/x-terminal/pull/23#discussion_r389630219

So you think the toggle should be on whether to insert or run the selection like terminus?

I still think more commands and different keymaps (<kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>r</kbd> and <kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>i</kbd>) would be more useful.

> the length of this discussion and the impassable impasse is disproportionate to the importance of the feature.

That may be true, but the discussion will be much easier to have now than in the future when someone wants to change the default.

Reply:
How about a toggle in the settings that enable a new command assigned to a default keybinding? Best of both worlds. Of course, more keybindings, but only for those who want it.

Reply:
>How about a toggle in the settings that enables "auto-move down" commands

We already discussed the toggle.https://github.com/bus-stop/x-terminal/pull/23#discussion_r389266825 and besides see https://github.com/bus-stop/x-terminal/pull/23#discussion_r392109193

Reply:
The toggle discussed so far was to enable the move-down commands *instead* of no-move-down-commands. My suggestion was to have all commands in there (and in the command palette) already, but only map the move-down ones keybindings if they are toggled on in the options. This way:

- users who don't need them don't enable them;
- users who need them but don't want to edit a keybinding file (which is admittedly a bit cumbersome in Atom) can just see and set everything they need from x-terminal settings, no risks of missing the feature and the toggle setting would be compatible with profiles
- users who need move-down commands but do not want to replace the no-move-down ones can have them all bound to keybinds, concurrently

This would mean @UziTech would still need to code a toggle, which he may think is overkill, but otherwise I believe that would clear all the limitations discussed above.

Reply:
equally solved by the toggle alone, the extra anything is just ott